ui/src/app/views/workspace-share/component.tsx

```diff
@@ -0,0 +1,508 @@
+    super(props);
+    this.state = {
+      autocompleteLoading: false,
+      autocompleteUsers: [],
+      userNotFound: '',
+      workspaceShareError: false,
+      usersLoading: false,
+      userRolesList: this.props.workspace.userRoles || [],
+      workspaceFound: (this.props.workspace !== null),
+      workspaceUpdateConflictError: false,
+      workspace: this.props.workspace,
+      searchTerm: '',
+      dropDown: false,
+    };
+    this.searchTermChangedEvent = fp.debounce(300, this.userSearch);
+  }
+
+  componentWillMount(): void {
+    document.addEventListener('mousedown', this.handleClickOutsideSearch, false);
+  }
+
+  componentWillUnmount(): void {
+    document.removeEventListener('mousedown', this.handleClickOutsideSearch, false);
+  }
+
+  save(): void {
+    if (this.state.usersLoading) {
+      return;
+    }
+    this.setState({usersLoading: true, workspaceShareError: false});
+    workspacesApi().shareWorkspace(this.state.workspace.namespace,
+      this.state.workspace.id,
+      {workspaceEtag: this.state.workspace.etag, items: this.state.workspace.userRoles})
+      .then((resp: ShareWorkspaceResponse) => {
+        const updatedWorkspace = this.state.workspace;
+        updatedWorkspace.etag = resp.workspaceEtag;
+        updatedWorkspace.userRoles = resp.items;
+        this.setState({usersLoading: false, userNotFound: '',
+          searchTerm: '', workspace: updatedWorkspace});
+        this.props.closeFunction();
+      }).catch(error => {
+        if (error.status === 400) {
+          this.setState({userNotFound: error});
+        } else if (error.status === 409) {
+          this.setState({workspaceUpdateConflictError: true});
+        } else {
+          this.setState({workspaceShareError: true});
+        }
+        this.setState({usersLoading: false});
+      });
+  }
+
+  removeCollaborator(user: UserRole): void {
+    this.setState(({userRolesList, workspace}) => (
+      {userRolesList: fp.remove(({email}) => user.email === email, userRolesList),
+        workspace: {...workspace,
+          userRoles: fp.remove(({email}) => {
+            return user.email === email;
+          }, workspace.userRoles)} as Workspace}
+    ));
+  }
+
+  reloadWorkspace(): void {
+    workspacesApi().getWorkspace(this.state.workspace.namespace, this.state.workspace.id)
+      .then((workspaceResponse) => {
+        this.setState({workspace: workspaceResponse.workspace});
+        this.resetModalState();
+      })
+      .catch(({status}) => {
+        if (status === 404) {
+          this.setState({workspaceFound: false});
+        }
+      });
+  }
+
+  addCollaborator(user: User): void {
+    const userRole: UserRole = {givenName: user.givenName, familyName: user.familyName,
+      email: user.email, role: WorkspaceAccessLevel.READER};
+    this.setState(({userRolesList, workspace}) => (
+      {searchTerm: '', autocompleteLoading: false, autocompleteUsers: [], dropDown: false,
```

Comment:
nit: sort of the same readability here, I feel like putting this bracket on the previous line is nicer